src/storage/src/source/sql_server/replication.rs

```diff
@@ -0,0 +1,337 @@
+                mz_sql_server_util::Client::connect(connection_config).await?;
+            // TODO(sql_server1): Move the connection into its own future.
+            mz_ore::task::spawn(|| "sql_server-connection", async move { connection.await });
+
+            let output_indexes: Vec<_> = outputs
+                .values()
+                .map(|v| usize::cast_from(v.partition_index))
+                .collect();
+
+            // Instances that have already made progress do not need to be snapshotted.
+            let needs_snapshot: BTreeSet<_> = outputs
+                .values()
+                .filter_map(|output| {
+                    if *output.resume_upper == [Lsn::minimum()] {
+                        Some(Arc::clone(&output.capture_instance))
+                    } else {
+                        None
+                    }
+                })
+                .collect();
+
+            // Map from a SQL Server 'capture instance' to Materialize collection.
+            let capture_instances: BTreeMap<_, _> = outputs
+                .values()
+                .map(|output| {
+                    (
+                        Arc::clone(&output.capture_instance),
+                        (output.partition_index, Arc::clone(&output.decoder)),
+                    )
+                })
+                .collect();
+            let mut cdc_handle = client.cdc(capture_instances.keys().cloned());
+
+            // Snapshot any instances that require it.
+            let snapshot_lsn = {
+                tracing::info!(?needs_snapshot, "starting snapshot");
+                let (snapshot_lsn, snapshot_streams) =
+                    cdc_handle.snapshot(Some(needs_snapshot)).await?;
+                let snapshot_cap = data_cap_set.delayed(&snapshot_lsn);
+
+                let mut snapshot_streams = std::pin::pin!(snapshot_streams);
+                while let Some((capture_instance, data)) = snapshot_streams.next().await {
+                    let sql_server_row = data.map_err(TransientError::from)?;
+
+                    // Decode the SQL Server row into an MZ one.
+                    let Some((partition_idx, decoder)) = capture_instances.get(&capture_instance)
+                    else {
+                        let definite_error = DefiniteError::ProgrammingError(format!(
+                            "capture instance didn't exist: '{capture_instance}'"
+                        ));
+                        let () = return_definite_error(
+                            definite_error,
+                            &output_indexes[..],
+                            data_output,
+                            data_cap_set,
+                            definite_error_handle,
+                            definite_error_cap_set,
+                        )
+                        .await;
+                        return Ok(());
+                    };
+
+                    // Failing to decode data is a permanent failure.
```

Comment:
It's not a permanent failure if the problematic row is later retracted so what we want to do here is emit the error in the output and continue as if nothing happened. The error message should contain enough data for the user to locate the problematic row (perhaps a raw representation)